import glob
import os

# Define input path
input_path = "/scratch/lustre/home/aust0075/Mouse_Aging_data"

samples = ["SRR5642553", "SRR5642554", "SRR5642555", "SRR5642545", "SRR5642544", "SRR5642561"]

# Reference genome files
genome_index_dir = "/scratch/lustre/home/aust0075/Mouse_Aging_data/grcm38/genome"
genome_gff = "/scratch/lustre/home/aust0075/Mouse_Aging_data/grmc38_gff"

# Rule all: final target files
rule all:
    input:
        expand("results/fastqc/{sample}_1_raw_fastqc.html", sample=samples),
        expand("results/fastqc/{sample}_2_raw_fastqc.html", sample=samples),
        expand("results/fastqc/{sample}_1_trimmed_fastqc.html", sample=samples),
        expand("results/fastqc/{sample}_2_trimmed_fastqc.html", sample=samples),
        expand("results/trim_galore/{sample}_1_trimmed.fq.gz", sample=samples),
        expand("results/trim_galore/{sample}_2_trimmed.fq.gz", sample=samples),
        expand("results/hisat2/{sample}.bam", sample=samples),
        "results/counts.txt",
        "results/multiqc_report.html"

rule fastqc:
    input:
        fq1 = lambda wildcards: f"{input_path}/{wildcards.sample}_1.fastq.gz",
        fq2 = lambda wildcards: f"{input_path}/{wildcards.sample}_2.fastq.gz"
    output:
        html1 = "results/fastqc/{sample}_1_raw_fastqc.html",
        html2 = "results/fastqc/{sample}_2_raw_fastqc.html",
        zip1 = "results/fastqc/{sample}_1_raw_fastqc.zip",
        zip2 = "results/fastqc/{sample}_2_raw_fastqc.zip"
    conda: "envs/preprocess_rnaseq.yaml"
    threads: 4
    shell:
        """
        fastqc -t {threads} --outdir results/fastqc {input.fq1} {input.fq2}
        """
rule fastqc_trimmed:
    input:
        fq1 = "results/trim_galore/{sample}_1_trimmed.fq.gz",
        fq2 = "results/trim_galore/{sample}_2_trimmed.fq.gz"
    output:
        html1 = "results/fastqc/{sample}_1_trimmed_fastqc.html",
        html2 = "results/fastqc/{sample}_2_trimmed_fastqc.html",
        zip1 = "results/fastqc/{sample}_1_trimmed_fastqc.zip",
        zip2 = "results/fastqc/{sample}_2_trimmed_fastqc.zip"
    conda: "envs/preprocess_rnaseq.yaml"
    threads: 4
    shell:
        """
        fastqc -t {threads} --outdir results/fastqc {input.fq1} {input.fq2}
        """


rule trimGalore:
    input:
        fq1 = lambda wildcards: f"{input_path}/{wildcards.sample}_1.fastq.gz",
        fq2 = lambda wildcards: f"{input_path}/{wildcards.sample}_2.fastq.gz"
    output:
        fq1_trimmed = "results/trim_galore/{sample}_1_trimmed.fq.gz",
        fq2_trimmed = "results/trim_galore/{sample}_2_trimmed.fq.gz"
    conda: "envs/preprocess_rnaseq.yaml"
    threads: 4
    params:
        quality = 15,
        length = 35
    shell:
        """
        trim_galore --paired -o results/trim_galore \
        -q {params.quality} \
        -j {threads} \
        --length {params.length} \
        {input.fq1} {input.fq2}
        """

rule hisat2:
    input:
        fq1 = "results/trim_galore/{sample}_1_trimmed.fq.gz",
        fq2 = "results/trim_galore/{sample}_2_trimmed.fq.gz"
    output:
        bam = "results/hisat2/{sample}.bam",
        bai = "results/hisat2/{sample}.bam.bai",
        stat = "results/hisat2/{sample}.stat",
        summary = "results/hisat2/{sample}_summary.txt"
    conda: "envs/preprocess_rnaseq.yaml"
    threads: 8
    shell:
        """
        hisat2 -p {threads} \
        -x {genome_index_dir} \
        -1 {input.fq1} -2 {input.fq2} \
        --summary-file {output.stat} | samtools view -bS - | samtools sort -o {output.bam}
        samtools index {output.bam}
        """

rule featureCounts:
    input:
        bam = expand("results/hisat2/{sample}.bam", sample=samples)
    output:
        counts = "results/counts.txt",
        stats = "results/counts.txt.summary"
    conda: "envs/preprocess_rnaseq.yaml"
    threads: 8
    shell:
        """
        featureCounts -a {genome_gff} -F GTF -t exon -g gene_id \
        -o {output.counts} -s 0 -T {threads} {input.bam}
        """

rule multiqc:
    input:
        fastqc_zip = expand("results/fastqc/{sample}_1_raw_fastqc.zip", sample=samples) +
                     expand("results/fastqc/{sample}_2_raw_fastqc.zip", sample=samples),
        trimmed_zip = expand("results/fastqc/{sample}_1_trimmed_fastqc.zip", sample=samples) +
                      expand("results/fastqc/{sample}_2_trimmed_fastqc.zip", sample=samples),
        alignment_summary = expand("results/hisat2/{sample}_summary.txt", sample=samples),
        counts_summary = "results/counts.txt.summary"
    output:
        html = "results/multiqc_report.html",
        data = directory("results/multiqc_data")
    log:
        "logs/multiqc.log"
    conda: "envs/preprocess_rnaseq.yaml"
    shell:
        """
        multiqc results/ \
        --ignore "results/hisat2/*/log/*" \
        -o results/ > {log} 2>&1
        """
